### PR TITLE
impl DoubleEndedIterator for {MatrixIter, MatrixIterMut}

### PR DESCRIPTION
**Pull request by Recmo**
Implements #794.

Also implements the [`FusedIterator`](https://doc.rust-lang.org/stable/core/iter/trait.FusedIterator.html) marker since it applies.

### CODE REVIEW

File: src/base/iter.rs [hunk 1]
```diff
@@ -119,6 +119,43 @@ macro_rules! iterator {
...
+                        debug_assert_eq!(outer_stride % inner_stride, 0);
+                        let num_rows = outer_stride / inner_stride;
+
+                        // Compute rows and cols remaining
+                        let outer_remaining = self.size / num_rows;
+                        let inner_remaining = self.size % num_rows;
+
+                        // Compute pointer to last element
+                        let last = self.ptr.offset(
+                            (outer_remaining * outer_stride + inner_remaining * inner_stride)
```
Comment (Recmo) [inline src/base/iter.rs:148]: Benchmark on my machine shows this to be 300x slower than forward iteration. Is there a way to simplify / optimize the math here?

File: src/base/iter.rs [hunk 2]
```diff
@@ -119,6 +119,43 @@ macro_rules! iterator {
...
+                        None
+                    } else {
+                        // Pre-decrement `size` such that it now counts to the
+                        // element we want to return.
+                        self.size -= 1;
+
+                        // Fetch strides
+                        let inner_stride = self.strides.0.value();
+                        let outer_stride = self.strides.1.value();
+                        debug_assert_eq!(outer_stride % inner_stride, 0);
```
Comment (Recmo) [inline src/base/iter.rs:139]: This assumption is probably false in MatrixSlice.

File: src/base/iter.rs (hunk 1 above)
Comment (Recmo) [reply src/base/iter.rs:149]: After the `inner_size` fix it is only 7x slower, which is acceptable to me.

**Comment by Recmo** [comment]
I'm not sure where the `check-fmt` failure comes from, the three errors are in files that this PR does not touch.

File: src/base/iter.rs [hunk 3]
```diff
@@ -111,6 +112,44 @@ macro_rules! iterator {
...
+                unsafe {
+                    if self.size == 0 {
+                        None
+                    } else {
+                        // Pre-decrement `size` such that it now counts to the
+                        // element we want to return.
+                        self.size -= 1;
+
+                        // Compute number of rows
+                        let inner_size = self.inner_end.offset_from(self.inner_ptr) as usize;
```
Comment (Recmo) [inline src/base/iter.rs:129]: This needs to be divided by `inner_size`

**Comment by sebcrozet** [comment]
Thank you for this PR! I took the liberty to rebase this on the head of `dev`.

Looks like the `cargo fmt` failures are gone now. I guess the failure was caused by the CI worker using a version of the compiler that changed some of the formats `cargo fmt` applies.

I'm waiting for the CI to complete before merging this now.

**Pull request merged**